AppActionVisualizer/app/src/main/java/com/example/appactionvisualizer/databean/AppActionsGenerator.java

```diff
@@ -14,41 +14,47 @@
 import java.util.Comparator;
 import java.util.HashMap;
 import java.util.HashSet;
-import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
-public class TestGenerator {
-  private static final String TAG = "TestGenerator";
-  public static List<AppAction> appActionsUnique = new ArrayList<>();
+public class AppActionsGenerator {
+  private static final String TAG = "AppActionsGenerator";
+  public static List<AppAction> appActions = new ArrayList<>();
   public static Map<ActionType, List<AppAction>> type2appActionList = new HashMap<>();
-  private static TestGenerator single_instance = null;
+  private static AppActionsGenerator single_instance = null;
 
-  public static TestGenerator getInstance() {
+  public static AppActionsGenerator getInstance() {
     if (single_instance == null)
-      single_instance = new TestGenerator();
+      single_instance = new AppActionsGenerator();
     return single_instance;
   }
 
   public void readFromFile(Context context) {
-    if(!appActionsUnique.isEmpty())
+    if(!appActions.isEmpty())
       return;
     InputStream is = context.getResources().openRawResource(R.raw.protobufbinary);
     try {
-      List<AppAction> appActions = new ArrayList<>();
       appActions.addAll(AppActionProtos.AppActions.parseFrom(is).getAppActionsList());
-      appActionsUnique = deduplication(appActions);
+      appActions = deduplication(appActions);
     } catch (IOException e) {
       e.printStackTrace();
     }
-    Collections.sort(appActionsUnique, new Comparator<AppAction>() {
+    sortAppActionByName(context, appActions);
+    parseDataToEachType(context, appActions);
+  }
+
+  private void sortAppActionByName(final Context context, final List<AppAction> appActions) {
+    Collections.sort(appActions, new Comparator<AppAction>() {
       @Override
       public int compare(AppAction t1, AppAction t2) {
-        return t2.getActionsCount() - t1.getActionsCount();
+        int strId1 = Utils.getResIdByPackageName(t1.getPackageName(), R.string.class);
```

Comment:
What is the sorting logic here? are we sorting by resId by alphabet order

Reply:
If any app action didn't find corresponding app name resource, it would have a lower priority; otherwise sort by the app name. Also comments added.

Reply:
Good question. Current built-in app icons and names are extracted from my phone through adb. So If new app action data is added, the new apps will have no icons and names if the user didn't install them manually. I'll refactor the codes to make this logic clearer.